contracts/utils/Swap.sol

```diff
@@ -0,0 +1,454 @@
+        if (bid.buyAmount < offer.minBidSize) {
+            errors[errCount] = "BID_TOO_SMALL";
+            errCount++;
+        }
+
+        // Check bid price
+        uint256 bidPrice =
+            (uint256(bid.sellAmount) *
+                uint256(10)**offer.offeredTokenDecimals) / bid.buyAmount;
+        if (bidPrice < offer.minPrice) {
+            errors[errCount] = "PRICE_TOO_LOW";
+            errCount++;
+        }
+
+        // Check signer allowance
+        uint256 signerAllowance =
+            IERC20(offer.biddingToken).allowance(
+                bid.signerWallet,
+                address(this)
+            );
+        if (signerAllowance < bid.sellAmount) {
+            errors[errCount] = "SIGNER_ALLOWANCE_LOW";
+            errCount++;
+        }
+
+        // Check signer balance
+        uint256 signerBalance =
+            IERC20(offer.biddingToken).balanceOf(bid.signerWallet);
+        if (signerBalance < bid.sellAmount) {
+            errors[errCount] = "SIGNER_BALANCE_LOW";
+            errCount++;
+        }
+
+        return (errCount, errors);
+    }
+
+    /**
+     * @notice Returns true if the nonce has been used
+     * @param signer address Address of the signer
+     * @param nonce uint256 Nonce being checked
+     */
+    function nonceUsed(address signer, uint256 nonce)
+        public
+        view
+        override
+        returns (bool)
+    {
+        uint256 groupKey = nonce / 256;
+        uint256 indexInGroup = nonce % 256;
+        return (_nonceGroups[signer][groupKey] >> indexInGroup) & 1 == 1;
+    }
+
+    /************************************************
+     *  INTERNAL FUNCTIONS
+     ***********************************************/
+
+    /**
+     * @notice Swap Atomic ERC20 Swap
+     * @param offer Offer struct containing offer details
+     * @param bid Bid struct containing bid details
+     */
+    function _swap(Offer storage offer, Bid calldata bid) internal {
+        require(DOMAIN_CHAIN_ID == getChainId(), "CHAIN_ID_CHANGED");
+
+        address signatory = _getSignatory(bid);
+        require(signatory != address(0), "SIGNATURE_INVALID");
+        require(signatory == bid.signerWallet, "SIGNATURE_MISMATCHED");
+        require(_markNonceAsUsed(signatory, bid.nonce), "NONCE_ALREADY_USED");
+        require(offer.availableSize > 0, "ZERO_AVAILABLE_SIZE");
+        require(bid.buyAmount >= offer.minBidSize, "BID_TOO_SMALL");
+        require(bid.buyAmount <= offer.totalSize, "BID_EXCEED_TOTAL");
+
+        // Ensure min. price is met
+        uint256 bidPrice =
+            (uint256(bid.sellAmount) *
+                uint256(10)**offer.offeredTokenDecimals) / bid.buyAmount;
+        require(bidPrice >= offer.minPrice, "PRICE_TOO_LOW");
+
+        // Check for partial fill
+        uint256 buyAmount = bid.buyAmount;
+        uint256 sellAmount = bid.sellAmount;
+        if (bid.buyAmount > offer.availableSize) {
+            buyAmount = offer.availableSize;
```

Comment:
Thinking about this - this behavior may be undesirable because we don't want to implicitly change the bid's buy amount to `offer.availableSize` if it exceeds it. It should just revert because we are exceeding the `availableSize`.